upstairs/src/lib.rs

```diff
@@ -2735,6 +2806,16 @@ impl Upstairs {
                 );
                 self.ds_transition(client_id, DsState::Deactivated);
                 self.set_inactive();
+            } else if err == CrucibleError::DecryptionError {
+                println!(
+                    "Authenticated decryption failed from client id {}!",
+                    client_id
+                );
+
+                // XXX reconciliation needs to occur, but do we trust that
+                // Downstairs anymore? once could imagine:
```

Comment:
Can we tell if this is malicious or just regular corruption?

Reply:
There is a distinction between an Err(...) response from the Downstairs, and an Ok(ReadResponse) where the data + nonce + tag does not decrypt properly. Using zfs should prevent the scenario where the read reports Ok but it's actually returned corrupted bytes.

Reaching this scenario should then mean either:

- an error has written junk Downstairs, or
- using the wrong key, or
- malicious tampering, or
- a corruption that zfs didn't catch, or
- something else?

Reply:
Using the wrong key is probably the most common one to run into.

Reply:
For the actual line 2816, Nit: "One could imagine"?  Or maybe you mean to say "Once upon a time?"

Reply:
I meant "one could imagine" haha, fixed in 1b9f56b